ckanext/data_qld/validation.py

```diff
@@ -21,16 +28,215 @@ def scheming_choices(field, schema):
+        else:
+            form_data = tk.request.params
+    except TypeError:
+        # working outside context, cli or tests
+        return
+
+    shema_upload = form_data.get(const.FIELD_SCHEMA_UPLOAD)
+
+    if is_uploaded_file(shema_upload):
+        schema_upload = form_data.get(const.FIELD_SCHEMA_UPLOAD)
+        return _get_underlying_file(schema_upload).read()
+
+
+def read_schema_from_file(data):
+    schema_upload_key = (const.FIELD_SCHEMA_UPLOAD, )
+    schema_upload = data.get(schema_upload_key)
+
+    if is_uploaded_file(schema_upload):
+        data[schema_upload_key] = ""
+        return _get_underlying_file(schema_upload).read()
+
+
+def get_schema_from_url(key, data, errors):
+    schema_url_key = (const.FIELD_SCHEMA_URL, )
+    value = data.get(schema_url_key)
+
+    if not value:
+        return
+
+    if (value and not isinstance(value, string_types)
+            or not is_url_valid(value)):
+        data[schema_url_key] = ""
+        err_msg = tk._('Must be a valid URL "{}"').format(value)
+        raise tk.Invalid(err_msg)
+
+    return value
+
+
+def get_schema_from_json(data):
+    schema_json_key = (const.FIELD_SCHEMA_JSON, )
+    value = data.get(schema_json_key)
+
+    if value:
+        data[schema_json_key] = ""
+        return value
+
+
+def _clear_pseudo_fields(data):
+    schema_json_key = (const.FIELD_SCHEMA_JSON, )
+    schema_url_key = (const.FIELD_SCHEMA_URL, )
+    schema_upload_key = (const.FIELD_SCHEMA_UPLOAD, )
+
+    data[schema_json_key] = data[schema_url_key] = data[schema_upload_key] = ""
+
+
+def align_default_schema(key, data, errors, context):
+    """Align resource schema with package schema"""
+    if not data[key]:
+        return
+
+    if context.get("ignore_auth"):
+        return
+
+    # key = ('resources', 0, 'align_default_schema')
+    if len(key) != 3 or key[2] != const.FIELD_ALIGNMENT:
+        return
+
+    default_schema = data[(const.FIELD_DEFAULT_SCHEMA, )]
+    idx = key[1]
+
+    resource_id = data.get((const.FIELD_RESOURCES, idx, 'id'))
+
+    if resource_id and _is_already_aligned(resource_id, default_schema,
+                                           context):
+        return
+
+    if not default_schema:
+        raise tk.Invalid(tk._("Missing {}").format(const.FIELD_DEFAULT_SCHEMA))
+
+    data[(const.FIELD_RESOURCES, idx, const.FIELD_RES_SCHEMA)] = default_schema
+
+
+def _is_already_aligned(resource_id, default_schema, context):
+    model = context['model']
+    session = context['session']
+
+    resource = session.query(model.Resource).get(resource_id)
+
+    if not resource:
+        return False
+
+    alignment_value = resource.extras.get(const.FIELD_ALIGNMENT)
+    resource_schema = resource.extras.get(const.FIELD_RES_SCHEMA)
+
+    if not resource_schema:
+        return False
+
+    if is_url_valid(resource_schema):
+        schemas_aligned = resource_schema == default_schema
+    else:
+        default_schema = json.loads(default_schema) if isinstance(
```

Comment:
This "parse as JSON else default" behaviour is used enough times that it should probably be extracted to a helper function.